File: l33tcode/max-sum-of-a-pair-with-equal-sum-of-digits.go
func maximumSum(nums []int) int {
	result := -1
	maxPrev := map[int]int{}

	for _, num := range nums {
		numCopy := num
		sumDigits := 0

		for numCopy > 0 {
			sumDigits = sumDigits + numCopy % 10
			numCopy = numCopy / 10
		}

		if prevNum, ok := maxPrev[sumDigits]; ok {
			if result < prevNum + num {
				result = prevNum + num
			}
			if prevNum < num {
				maxPrev[sumDigits] = num
			}
		} else {
			maxPrev[sumDigits] = num
		}
	}

	return result
}

File: l33tcode/max-sum-of-a-pair-with-equal-sum-of-digits.py
class Solution:
    def maximumSum(self, nums: list[int]) -> int:
        result = -1
        max_prev: dict[int, int] = {}

        for num in nums:
            sum_digits = 0
            num_copy = num
            while num_copy:
                sum_digits += num_copy % 10
                num_copy //= 10

            if prev_num := max_prev.get(sum_digits):
                result = max(result, prev_num + num)

            max_prev[sum_digits] = max(max_prev.get(sum_digits, 0), num)

        return result
